src/Formats/CapnProtoSerializer.cpp

```diff
@@ -336,23 +375,43 @@ namespace
     private:
         UInt16 getValue(const ColumnPtr & column, size_t row_num)
         {
+            if (!capnp_enum_is_superset)
+                throw Exception(ErrorCodes::CAPN_PROTO_BAD_CAST, "Cannot convert ClickHouse enum to CapnProto enum: CapnProto enum values/names is not a superset of ClickHouse enum values/names");
+
             EnumType enum_value = assert_cast<const ColumnVector<EnumType> &>(*column).getElement(row_num);
             if (enum_comparing_mode == FormatSettings::CapnProtoEnumComparingMode::BY_VALUES)
                 return static_cast<UInt16>(enum_value);
-            return ch_to_capnp_values[enum_value];
+            auto it = ch_to_capnp_values.find(enum_value);
+            if (it == ch_to_capnp_values.end())
+                throw Exception(ErrorCodes::INCORRECT_DATA, "Unexpected value {} in ClickHouse enum", enum_value);
+
+            return it->second;
         }
 
         void insertValue(IColumn & column, UInt16 capnp_enum_value)
         {
+            if (!ch_enum_is_superset)
+                throw Exception(ErrorCodes::CAPN_PROTO_BAD_CAST, "Cannot convert CapnProto enum to ClickHouse enum: ClickHouse enum values/names is not a superset of CapnProto enum values/names");
+
             if (enum_comparing_mode == FormatSettings::CapnProtoEnumComparingMode::BY_VALUES)
+            {
                 assert_cast<ColumnVector<EnumType> &>(column).insertValue(static_cast<EnumType>(capnp_enum_value));
+            }
             else
+            {
+                auto it = capnp_to_ch_values.find(capnp_enum_value);
+                if (it == capnp_to_ch_values.end())
+                    throw Exception(ErrorCodes::INCORRECT_DATA, "Unexpected value {} in CapnProto enum", capnp_enum_value);
+
                 assert_cast<ColumnVector<EnumType> &>(column).insertValue(capnp_to_ch_values[capnp_enum_value]);
```

Comment:
s/`capnp_to_ch_values[capnp_enum_value]`/`it->second`/

Reply:
Oh, nice catch!